### PR TITLE
refine class member function implementation in ProtocolV2 

### PR DESCRIPTION
**Pull request by changchengx**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/msg/async/ProtocolV2.cc [hunk 1]
```diff
@@ -743,7 +743,8 @@ CtPtr ProtocolV2::read(CONTINUATION_RXBPTR_TYPE<ProtocolV2> &next,
   next.node = std::move(buffer);
   ssize_t r = connection->read(len, buf,
     [&next, this](char *buffer, int r) {
-      if (unlikely(pre_auth.enabled) && r >= 0) {
+      ceph_assert(r > 0);
```
Comment (tchaikov) [inline src/msg/async/ProtocolV2.cc:746]: i'd prefer s/ceph_assert/assert/ in the critical path.

**Comment by changchengx** [comment]
Sorry for the late reply. I haven't received email notification.

**Comment by changchengx** [comment]
test this please

File: src/msg/async/ProtocolV2.cc [hunk 2]
```diff
@@ -743,7 +743,8 @@ CtPtr ProtocolV2::read(CONTINUATION_RXBPTR_TYPE<ProtocolV2> &next,
   next.node = std::move(buffer);
   ssize_t r = connection->read(len, buf,
     [&next, this](char *buffer, int r) {
-      if (unlikely(pre_auth.enabled) && r >= 0) {
+      assert(r > 0);
```
Comment (tchaikov) [inline src/msg/async/ProtocolV2.cc:746]: what is this change for?
```c++
    case STATE_CONNECTION_ESTABLISHED: {
      if (pendingReadLen) {
        ssize_t r = read(*pendingReadLen, read_buffer, readCallback);
        if (r <= 0) { // read all bytes, or an error occured
          pendingReadLen.reset();
          char *buf_tmp = read_buffer;
          read_buffer = nullptr;
          readCallback(buf_tmp, r);
        }
        logger->tinc(l_msgr_running_recv_time,
            ceph::mono_clock::now() - recv_start_time);
        return;
      }
```

i found this in `AsyncConnection::process()`. if we have `assert(r > 0)`, will the program abort?

**Comment by changchengx** [comment]
Abandon.
I have no time to refine this PR.

**Comment by tchaikov** [comment]
@changchengx could you keep this PR around ? i think it is still relevant . i just need more time to review it

**Comment by changchengx** [comment]
@tchaikov 
I'll open this PR. After finishing refining it, I'll ping you to review it.
Thanks for your right suggestion.

**Comment by tchaikov** [comment]
> @tchaikov
> I'll open this PR. After finishing refining it, I'll ping you to review it.
> Thanks for your right suggestion.

i dropped your PR of "msg: correct check main_preamble.num_segments". as it does not apply anymore.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-07-29_04:08:59-rados-wip-kefu-testing-2020-07-29-1027-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-29_11:42:32-rados-wip-kefu-testing-2020-07-29-1722-distro-basic-smithi/

**Pull request merged**